Refactor Task 1,2,3. Finished Task number 4,5,7 added 2 properties to…

zur250 opened the pull request:
… the activity : min_time_to_start and max_time_to_start. Added 2 new methods to the project : calc activity min time and calc activity max time. finished calc activity min time.

zur250 committed:
Refactor Task 1,2,3. Finished Task number 4,5,7 added 2 properties to the activity : min_time_to_start and max_time_to_start. Added 2 new methods to the project : calc activity min time and calc activity max time. finished calc activity min time.

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
+                    dep_activity.max_time_to_start = dep_activity.min_time_to_start
+
+    def calculate_activity_max_time_to_start(self):
+        '''for activity, depending_activity in self._project.items():
```
why you have the ''' at the start and at the end ?

zur250 replied:
it is a comment for multiple line. its not good yet

OrenZak replied:
👍

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
+        print("Activity %s with duration %s", self.name, self.duration)
+
+    def __eq__(self, other):
+        return self.name == other.name and self.duration == other.duration
```
not sure about it, maybe we should generate a UUID to each activity on create time and just compare it.

zur250 replied:
it might be a good idea.

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
+        else:
+            self.__project = {}
+
+        self.__project_duration = 0
```
`__project_duration` should be calculated if there is a project content.

zur250 replied:
i will midify it tomorrow

zur250 replied:
done

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
+                self._project[activity].append(depending_activity)
+                logger.debug("%s added to the activity", depending_activity)
+            else:
+                logger.debug("%s already in the list of depending activities", depending_activity)
```
maybe we should update, in case the activity content was changed

zur250 replied:
it suppose to check the equals method that i implemented on the activity class.

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
+            self._project[activity] = []
+            logger.debug("%s added to the project", activity)
+        else:
+            logger.debug("%s already at the", activity)
```
also here maybe we should update the activity

zur250 replied:
it suppose to check the equals method that i implemented on the activity class.

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
-        if activity in self._project:
-            self._project.pop(activity)
+        # this method removing activity from the project
+        if activity in self.__project:
```
lets log the `else` option: `logger.debug("%s trying to delete activity that not in the project", activity)`

zur250 replied:
will modify tomorrow

zur250 replied:
done

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
+
+    def calculate_activity_max_time_to_start(self):
+        '''for activity, depending_activity in self._project.items():
+            for dep_activity in depending_activity:
```
rename second param to `depending_activities`

zur250 replied:
will modify

zur250 replied:
done

OrenZak commented on Cpm.py:
```diff
@@ -4,43 +4,98 @@
-        # todo find critical path of the project (Showing the edges of the critical pass with their length
-        pass
+        # find critical path of the project (Showing the edges of the critical pass with their length)
+        for activity in self.__project.keys():
```
how do we make sure it starts form the first created activity?

zur250 replied:
idk

OrenZak reviewed (approved):
see my comments 👍

zur250 committed:
Changing all CPM buisness logic.
Change there are 3 classes :
Activity - this is the arrows.
Bullets - this is te circles.
Project - this is the cpm manager
TODO - create a test CPM project and start debug all the methods.
Finish the __str__ method and the slacks method.
Verify its working

OrenZak reviewed (approved):
👍
